File: java-advanced/src/main/java/functions/utils/Logger.java
package functions.utils;

import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;

public class Logger {
    public enum Level {
        INFO, ERROR, WARN
    }

    private void log(Level level, String message, String meta) {
        String timestamp = LocalDateTime.now().format(DateTimeFormatter.ISO_DATE_TIME);
        String logEntry = String.format(
            "{\"timestamp\":\"%s\",\"level\":\"%s\",\"message\":\"%s\"%s}",
            timestamp, level, message, meta.isEmpty() ? "" : ",\"meta\":" + meta
        );
        System.out.println(logEntry);
    }

    public void info(String message) {
        log(Level.INFO, message, "");
    }

    public void info(String message, String meta) {
        log(Level.INFO, message, meta);
    }

    public void error(String message) {
        log(Level.ERROR, message, "");
    }

    public void error(String message, String meta) {
        log(Level.ERROR, message, meta);
    }

    public void warn(String message) {
        log(Level.WARN, message, "");
    }

    public void warn(String message, String meta) {
        log(Level.WARN, message, meta);
    }
}

File: nodejs-advanced/src/utils/logger.js
const logLevels = {
  INFO: 'INFO',
  ERROR: 'ERROR',
  WARN: 'WARN'
};

function log(level, message, meta = {}) {
  const timestamp = new Date().toISOString();
  console.log(JSON.stringify({
    timestamp,
    level,
    message,
    ...meta
  }));
}

module.exports = {
  info: (message, meta) => log(logLevels.INFO, message, meta),
  error: (message, meta) => log(logLevels.ERROR, message, meta),
  warn: (message, meta) => log(logLevels.WARN, message, meta)
};
